### PR TITLE
added change to iterate through all permuatations

### PR DESCRIPTION
**Pull request by wengshian94**
Added change for multiple beta reduction paths.

### CODE REVIEW

File: lambda_calc/wrapper.py [hunk 1]
```diff
@@ -38,13 +39,32 @@ def store_all_reducs(self):
         current_expr = self.expr
         while to_go:
             try:
-                curr_reduced = next(all_beta_reductions(current_expr))
-                reducs_list.append(curr_reduced)
-                current_expr = curr_reduced
-            except:
+                reduced = all_beta_reductions(current_expr)
+                curr_reduced = self.collapse_reductions(reduced)
```
Comment (zojize) [inline lambda_calc/wrapper.py]: why not just `list(reduced)`

File: lambda_calc/wrapper.py [hunk 2]
```diff
@@ -29,6 +29,7 @@ def __init__(self, lambdastring : str, candidatestring : str):
         #Check the beta reductions
         self.check_all_beta_reducs()
 
+
     def store_all_reducs(self):
```
Comment (zojize) [inline lambda_calc/wrapper.py]: I feel like there's no need to store all reductions beforehand? especially if we're iterating through all reductions. Why not just compute the reductions of the previous step to verify that a step is valid?

File: lambda_calc/wrapper.py [hunk 3]
```diff
@@ -38,13 +39,32 @@ def store_all_reducs(self):
...
+        """
+        Collapse all possible beta reductions into a list
+        """
+        reduc_list = []
+        to_add = reductions
+        while True:
+            try:
+                to_add = next(to_add)
+                reduc_list.append(to_add)
+            except Exception as e:
```
Comment (zojize) [inline lambda_calc/wrapper.py]: error handling should be as explicit as possible, i.e. StopIteration here. But this function could've just been replaced by `list`.

**Review by zojize — COMMENTED** [review]
made some comments, please take a look.
also, adding to the oop conversation in the previous PR

> I think it should be a class since a lot of variables and expressions are stored

I don't think oop has a clear advantage here, the use of class often is to help manage persistent states and help describe data that has an obvious inheritance structure. There is no need for persistent state since the whole thing will only run once per submission, and I see no reason why you might want to inherit from this class. All class variables can just be declare inside the function as local variables and I think its more clear since its one less layer of abstraction and you don't need `self.xxx`. It also 

>  Also, for the Prairelearn check, it can be easily used to provide hints without having to explicitly store variables

I'm not sure what you mean here, if you're talking about the aggregated error messages, they could still be returned at the end of the function.

> The use of OOP might also make the code cleaner and also makes it easier to hook it up with Mattox's code.

Well since Mattox is an FP guy he might like functions more? I don't know, this is up to him.

Basically my rule of thumb is, I would choose to not use classes unless there is something that can only be done using classes. It's all about communication at the end of the day, if oop can't help express the code better or create a more concise API, I would rather not use it.

File: lambda_calc/wrapper.py (hunk 2 above)
Comment (wengshian94) [reply lambda_calc/wrapper.py]: Hmm then how would you know the difference between a wrong logical follow-up from the beta-reduction and an unnecessary line?

Comment (wengshian94) [reply lambda_calc/wrapper.py]: I think we could just treat them the same in this case

File: lambda_calc/parser.py [hunk 4]
```diff
@@ -10,7 +10,7 @@ def lexeme(p: Parser) -> Parser: return p << ws
 
 
 lambda_ = lexeme(regex(r'[λ\\L]')).desc('lambda')
-var = lexeme(regex(r"[a-z]")).map(Var).desc('variable')
+var = lexeme(regex(r"[a-z](?:')?|[a-z]''")).map(Var).desc('variable')
```
Comment (zojize) [inline lambda_calc/parser.py]: even though it's unlikely that there will be more than 2 `'`s but let's just do `r'[a-z]\'*'` to make the parser more forgiving. everything else looks good!

Comment (wengshian94) [reply lambda_calc/parser.py]: sounds good, done
